From 5e80423fd37f3a9ed663c29f5534d56abc464ae2 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Mon, 5 Aug 2024 15:37:43 +0800
Subject: [PATCH 93/97] mtk: mac80211: fix wds channel switch crash issue

The link->reserved of the AP VLAN might be used in the following commit
change.
However, the link->reserved of the AP VLAN is not initialized and not
assigned during the channel switch, so we should use the link->reserved
of the AP instead.
Without this fix, wds ap & sta will crash during the channel switch.
https://gerrit.mediatek.inc/plugins/gitiles/gateway/WiFi/mtk-wireless-mac80211/+/b27512368591fc959768df1f7dacf2a96b1bd036

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 net/mac80211/chan.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/net/mac80211/chan.c b/net/mac80211/chan.c
index 035a309..0532470 100644
--- a/net/mac80211/chan.c
+++ b/net/mac80211/chan.c
@@ -452,6 +452,16 @@ static void ieee80211_chan_bw_change(struct ieee80211_local *local,
 			else
 				new_chandef = &link_conf->chanreq.oper;
 
+			/* Access the reserved chanreq of the AP when it is AP VLAN */
+			if (reserved && sdata->vif.type == NL80211_IFTYPE_AP_VLAN) {
+				struct ieee80211_sub_if_data *ap;
+				struct ieee80211_link_data *ap_link;
+
+				ap = container_of(sdata->bss, struct ieee80211_sub_if_data, u.ap);
+				ap_link = rcu_dereference(ap->link[link_id]);
+				new_chandef = &ap_link->reserved.oper;
+			}
+
 			new_sta_bw = _ieee80211_sta_cur_vht_bw(link_sta,
 							       new_chandef);
 
-- 
2.18.0

